### PR TITLE
Log message when queue has everyone/anonymous logon permissions

### PR DESCRIPTION
**Pull request by hmemcpy**
This adds a warning when creating/starting the queue, if it has permissions for `Everyone` and `Anonymous` users.

### CODE REVIEW

File: src/NServiceBus.Core/Transports/Msmq/Config/MsmqTransportConfigurator.cs [hunk 1]
```diff
@@ -18,6 +22,54 @@ public class MsmqTransportConfigurator : ConfigureTransport
...
+            void WarnIfPublicAccess(MessageQueue queue)
+            {
+                MessageQueueAccessRights? everyoneRights, anonymousRights;
+
+                queue.TryGetPermissions(MsmqConstants.LocalAnonymousLogonName, out anonymousRights);
+                queue.TryGetPermissions(MsmqConstants.LocalEveryoneGroupName, out everyoneRights);
+
+                if (anonymousRights.HasValue && everyoneRights.HasValue)
+                {
+                    Logger.WarnFormat("Queue [{0}] is running with [{1}] and [{2}] permissions. Consider changing those, as required",
```
Comment (johnsimons) [inline src/NServiceBus.Core/Transports/Msmq/Config/MsmqTransportConfigurator.cs]: We should give the user a link to best practices ?

File: src/NServiceBus.Core/Transports/Msmq/MsmqQueueCreator.cs [hunk 2]
```diff
@@ -92,6 +96,7 @@ static void CreateQueue(string queuePath, string account, bool transactional)
             }
 
             Logger.DebugFormat("Created queue, path: [{0}], account: [{1}], transactional: [{2}]", queuePath, account, transactional);
+            Logger.WarnFormat("Queue created with [{0}] and [{1}] permissions. Consider changing those, as required", MsmqConstants.LocalEveryoneGroupName, MsmqConstants.LocalAnonymousLogonName);
```
Comment (johnsimons) [inline src/NServiceBus.Core/Transports/Msmq/MsmqQueueCreator.cs]: This seems weird to me, so we are saying that `Everyone` is probably not the right ACL, but we are still adding it by default, so there is no way for the user to avoid this warning!

Comment (SimonCropp) [inline src/NServiceBus.Core/Transports/Msmq/MsmqQueueCreator.cs]: @johnsimons thats why the issues says 

> Read the permissions at startup. if anon and everyone is enabled then write to the log. Info if the debugger is attached; otherwise warn

https://github.com/Particular/FeatureDevelopment/issues/303

File: src/NServiceBus.Core/Transports/Msmq/MsmqQueueCreator.cs [hunk 3]
```diff
@@ -7,12 +7,16 @@ namespace NServiceBus.Transports.Msmq
     using Logging;
     using Support;
 
+    static class MsmqConstants
```
Comment (SimonCropp) [inline src/NServiceBus.Core/Transports/Msmq/MsmqQueueCreator.cs]: why the new class?

File: src/NServiceBus.Core/Utils/MsmqExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        public static bool TryGetPermissions(this MessageQueue queue, string user, out MessageQueueAccessRights? rights)
+        {
+            string sid = GetSidForUser(user);
+
+            try
+            {
+                rights = GetPermissions(queue.FormatName, sid);
+                return true;
+            }
+            catch
```
Comment (SimonCropp) [inline src/NServiceBus.Core/Utils/MsmqExtensions.cs]: no value in catching the specific exception?

File: src/NServiceBus.Core/Transports/Msmq/MsmqQueueCreator.cs (hunk 3 above)
Comment (hmemcpy) [inline src/NServiceBus.Core/Transports/Msmq/MsmqQueueCreator.cs]: @SimonCropp I wanted to reuse those constants elsewhere, and making them internal within `QueueCreator` didn't feel right.

File: src/NServiceBus.Core/Utils/MsmqExtensions.cs (hunk 4 above)
Comment (hmemcpy) [inline src/NServiceBus.Core/Utils/MsmqExtensions.cs]: @SimonCropp Possibly, I need to go over this code again to comment it a bit (most of it taken from referencesource of `MessageQueue` class, some from StackOverflow). There were some exceptions thrown that were internal in the framework, so I threw a plain old Exception instead. I'll fix it.

File: src/NServiceBus.Core/Transports/Msmq/Config/MsmqTransportConfigurator.cs (hunk 1 above)
Comment (hmemcpy) [inline src/NServiceBus.Core/Transports/Msmq/Config/MsmqTransportConfigurator.cs]: @johnsimons good idea, one of the tasks here is fix the docs, but I think I read somewhere we don't like to use links in messages, because they might become 404?

Comment (SzymonPobiega) [inline src/NServiceBus.Core/Transports/Msmq/Config/MsmqTransportConfigurator.cs]: +1 for more descriptive message
-1 for link to the docs as it might break

File: src/NServiceBus.Core/Utils/MsmqExtensions.cs (hunk 4 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Utils/MsmqExtensions.cs]: Do we want to set a threshold of how much 'hacks' can this feature require? I mean, it is super cool that @hmemcpy found this solution but isn't it a bit of a problem in terms of maintainability?

Comment (hmemcpy) [inline src/NServiceBus.Core/Utils/MsmqExtensions.cs]: @SzymonPobiega yes, I'm adding some comments explaining this, but it's definitely bit of an overkill. Still, a way to read permissions on an existing queue is a nice thing to have :)

Comment (johnsimons) [inline src/NServiceBus.Core/Utils/MsmqExtensions.cs]: That is not a bad idea at all @hmemcpy, a separate tool to provide a report on the MSMQ queue ACLS would in my opinion be a better option.
This logging things onto a file is in my opinion a bit too optimistic.

File: src/NServiceBus.Core/Utils/MsmqExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    using System.Security.Principal;
+
+    /// <summary>
+    /// Reads the Access Control Entries (ACE) from an MSMQ queue.
+    /// </summary>
+    /// <remarks>
+    /// There is no managed API for reading the queue permissions, this has to be done via P/Invoke. by calling <c>MQGetQueueSecurity</c> API.
+    /// See http://stackoverflow.com/questions/10177255/how-to-get-the-current-permissions-for-an-msmq-private-queue
+    /// </remarks>
+    static class MsmqExtensions
```
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Utils/MsmqExtensions.cs]: Should this be `MessageQueueExtensions` ? (nitpicking)

Comment (hmemcpy) [inline src/NServiceBus.Core/Utils/MsmqExtensions.cs]: @SzymonPobiega good point, I named it before realizing what I'm dealing with

**Comment by SzymonPobiega** [comment]
:+1: for merging

**Comment by SzymonPobiega** [comment]
@SimonCropp as you initiated that thing, can you review the implementation?

**Comment by hmemcpy** [comment]
I'm ready for my closeup @Particular/nservicebus ...

File: src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            Assert.IsNotNull(logItem);
+            StringAssert.Contains(@"is running with [Everyone] and [NT AUTHORITY\ANONYMOUS LOGON] permissions. Consider setting appropriate permissions, if required by your organization", logItem.Message);
+        }
+
+        public class Context : ScenarioContext
+        {
+            public bool WasCalled { get; set; }
+            public Guid Id { get; set; }
+
+            public bool GeneratorWasCalled { get; set; }
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs]: This is not used

File: src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                    return;
+                }
+                initialized = true;
+                EndpointSetup<DefaultServer>(c =>
+                {
+                    c.UseTransport<MsmqTransport>();
+                });
+            }
+
+            static Context Context { get; set; }
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs]: Why is this needed?

File: src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+        public class Context : ScenarioContext
+        {
+            public bool WasCalled { get; set; }
+            public Guid Id { get; set; }
+
+            public bool GeneratorWasCalled { get; set; }
+        }
+
+        public class EndPoint : EndpointConfigurationBuilder, IWantToRunBeforeConfigurationIsFinalized
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs]: `IWantToRunBeforeConfigurationIsFinalized` why do you need this?

Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs]: Don't you think a default endpoint would be enough?

File: src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        public class MyMessage : ICommand
+        {
+            public Guid Id { get; set; }
+        }
+
+        public class MyMessageHandler : IHandleMessages<MyMessage>
+        {
+            public Context Context { get; set; }
+
+            public IBus Bus { get; set; }
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs]: Remove if not needed

File: src/NServiceBus.Core.Tests/Msmq/MsmqExtensionsTests.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        static string LocalEveryoneGroupName = new SecurityIdentifier(WellKnownSidType.WorldSid, null).Translate(typeof(NTAccount)).ToString();
+        static string LocalAnonymousLogonName = new SecurityIdentifier(WellKnownSidType.AnonymousSid, null).Translate(typeof(NTAccount)).ToString();
+
+        string path;
+        MessageQueue queue;
+
+        [TestFixtureSetUp]
+        public void Setup()
+        {
+            var queueName = "labelTest";
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Msmq/MsmqExtensionsTests.cs]: Interesting name

File: src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs (hunk 8 above)
Comment (hmemcpy) [inline src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs]: I... don't know, I used another test as a basis of my own. I ran into some problems, and so I decided to just copy it for now...

File: src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs (hunk 7 above)
Comment (hmemcpy) [inline src/NServiceBus.AcceptanceTests/Basic/When_starting_up_the_endpoint.cs]: @danielmarbach honestly, I used another test as a basis for this, because when I tried to do it by hand (e.g. without sending an actual message, or defining a handler), the test wouldn't finish. So I just used clipboard inheritance to fix it :)

File: src/NServiceBus.Core.Tests/Msmq/MsmqTestsBase.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
+﻿namespace NServiceBus.Core.Tests.Msmq
+{
+    using System.Messaging;
+
+    public class MsmqTestsBase
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/Msmq/MsmqTestsBase.cs]: I don't see a lot of value in this base class. To be frank I'm not a big fan of base classes. When we keep it we should make it a bit more useful and potentially rename it `SingleMsmqQueueTestBase` and add the setup and the teardown of the queue to the base. I.ex. inheritors could override an abstract property called `QueueName` or something

File: src/NServiceBus.Core/Transports/Msmq/MsmqConstants.cs [hunk 12]
```diff
@@ -0,0 +1,11 @@
+namespace NServiceBus.Transports.Msmq
+{
+    using System.Security.Principal;
+
+    static class MsmqConstants
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Transports/Msmq/MsmqConstants.cs]: Nitpicker: If we are not really reusing these constants I somehow agree with @SimonCropp that we could add it back to the `MsmqQueueCreator`.

File: src/NServiceBus.Core.Tests/Msmq/MsmqTestsBase.cs (hunk 11 above)
Comment (hmemcpy) [inline src/NServiceBus.Core.Tests/Msmq/MsmqTestsBase.cs]: @danielmarbach I agree, but I was looking for a test that did a similar thing, and this time I *didn't* want to copypaste all the setup code to my own, so I extracted this base class. Could be better, I agree.

File: src/NServiceBus.Core/Transports/Msmq/MsmqConstants.cs (hunk 12 above)
Comment (hmemcpy) [inline src/NServiceBus.Core/Transports/Msmq/MsmqConstants.cs]: @danielmarbach ok, I thought about reusing those, but I ended up not to. I'll revert this.

File: src/NServiceBus.Core.Tests/Msmq/MsmqTestsBase.cs (hunk 11 above)
Comment (SimonCropp) [inline src/NServiceBus.Core.Tests/Msmq/MsmqTestsBase.cs]: kill off the base class and use helper methods instead

**Comment by hmemcpy** [comment]
All good now @danielmarbach @SimonCropp 
image

**Comment by SimonCropp** [comment]
looks good

**Comment by hmemcpy** [comment]
Can any one of you please merge it?
